homeassistant/components/vesync/fan.py

```diff
@@ -0,0 +1,117 @@
+"""Support for VeSync fans."""
+import logging
+
+from homeassistant.components.fan import (
+    SPEED_HIGH,
+    SPEED_LOW,
+    SPEED_MEDIUM,
+    SPEED_OFF,
+    SUPPORT_SET_SPEED,
+    FanEntity,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from .common import VeSyncDevice
+from .const import DOMAIN, VS_DISCOVERY, VS_DISPATCHERS, VS_FANS
+
+_LOGGER = logging.getLogger(__name__)
+
+DEV_TYPE_TO_HA = {
+    "LV-PUR131S": "fan",
+}
+
+SPEED_AUTO = "auto"
+FAN_SPEEDS = [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the VeSync fan platform."""
+
+    async def async_discover(devices):
+        """Add new devices to platform."""
+        _async_setup_entities(devices, async_add_entities)
+
+    disp = async_dispatcher_connect(hass, VS_DISCOVERY.format(VS_FANS), async_discover)
+    hass.data[DOMAIN][VS_DISPATCHERS].append(disp)
+
+    _async_setup_entities(hass.data[DOMAIN][VS_FANS], async_add_entities)
+    return True
+
+
+@callback
+def _async_setup_entities(devices, async_add_entities):
+    """Check if device is online and add entity."""
+    dev_list = []
+    for dev in devices:
+        if DEV_TYPE_TO_HA.get(dev.device_type) == "fan":
+            dev_list.append(VeSyncFanHA(dev))
+        else:
+            _LOGGER.warning(
+                "%s - Unknown device type - %s", dev.device_name, dev.device_type
+            )
+            continue
+
+    async_add_entities(dev_list, update_before_add=True)
+
+
+class VeSyncFanHA(VeSyncDevice, FanEntity):
+    """Representation of a VeSync fan."""
+
+    def __init__(self, fan):
+        """Initialize the VeSync fan device."""
+        super().__init__(fan)
+        self.smartfan = fan
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return SUPPORT_SET_SPEED
+
+    @property
+    def speed(self):
+        """Return the current speed."""
+        if self.smartfan.mode == SPEED_AUTO:
+            return SPEED_AUTO
```

Comment:
Since the fan can still return `SPEED_AUTO`, and we have taken it out of `FAN_SPEEDS` this puts us in an invalid state.  Unfortunately it doesn't look like we can tell what speed the fan is actually running it so it leaves us in a position where violating the spec might make more sense since auto can be set from outside Home Assistant. 

I think its ok to add back in as long as we refactor after https://github.com/home-assistant/architecture/issues/127 resolves this.

Reply:
Sounds good and yes I will refactor when a decision is made.

Reply:
It's not ok to violate our architecture design until the architecture has been changed. Please remove the auto speed.